storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,766 @@
+        assert(dynamicState.container != null);
+        assert(dynamicState.currentAssignment != null);
+        
+        dynamicState.container.cleanUp();
+        staticState.localizer.releaseSlotFor(dynamicState.currentAssignment.get_topology_id(), staticState.port);
+        DynamicState ret = dynamicState.withCurrentAssignment(null, null);
+        if (nextState != null) {
+            ret = ret.withState(nextState);
+        }
+        return ret;
+    }
+    
+    /**
+     * State Transitions for WAITING_FOR_BLOB_LOCALIZATION state.
+     * PRECONDITION: neither pendingLocalization nor pendingDownload is null.
+     * PRECONDITION: The slot should be empty
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception on any error
+     */
+    static DynamicState handleWaitingForBlobLocalization(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.pendingLocalization != null);
+        assert(dynamicState.pendingDownload != null);
+        assert(dynamicState.container == null);
+        
+        try {
+            dynamicState.pendingDownload.get(1000, TimeUnit.MILLISECONDS);
+            //Downloading of all blobs finished.
+            if (!equivalent(dynamicState.newAssignment, dynamicState.pendingLocalization)) {
+                //Scheduling changed
+                staticState.localizer.releaseSlotFor(dynamicState.pendingLocalization.get_topology_id(), staticState.port);
+                return prepareForNewAssignmentOnEmptySlot(dynamicState, staticState);
+            }
+            Container c = staticState.containerLauncher.launchContainer(staticState.port, dynamicState.pendingLocalization, staticState.localState);
+            return dynamicState.withCurrentAssignment(c, dynamicState.pendingLocalization).withState(MachineState.WAITING_FOR_WORKER_START).withPendingLocalization(null, null);
+        } catch (TimeoutException e) {
+            //We waited for 1 second loop around and try again....
+            return dynamicState;
+        }
+    }
+    
+    /**
+     * State Transitions for WAITING_FOR_BASIC_LOCALIZATION state.
+     * PRECONDITION: neither pendingLocalization nor pendingDownload is null.
+     * PRECONDITION: The slot should be empty
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception on any error
+     */
+    static DynamicState handleWaitingForBasicLocalization(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.pendingLocalization != null);
+        assert(dynamicState.pendingDownload != null);
+        assert(dynamicState.container == null);
+        
+        //Ignore changes to scheduling while downloading the topology code
```

Comment:
Might want to add the reasoning: It is because properly cancelling the other thread's downloads while other workers might need the same files adds a ton of complexity for an uncommon case.